### ISSUE TITLE
Add Adabelief optimizer to the C++ API

### ISSUE THREAD

**Issue by matovitch** [pytorch/pytorch#47571]
## 🚀 Feature
Add Adabelief optimizer to the C++ API

## Motivation
I have been using this new optimizer recently with the Python API and found out it worked extremely well (for my use cases at least). I also find the paper and theory behind it pretty neat. I would be ready to work on it (although with no guarantee since this would be done on my free time) if you consider this is worth adding to the API.

Thank you for reading this far. Cheers. :)

Link to the paper: https://arxiv.org/abs/2010.07468

cc @albanD @mruberry @jbschlosser @vincentqb @yf225 @glaringlee

**Comment by glaringlee** [comment]
@matovitch 
Thanks for reaching out. I think this is a very nice feature, I also find other people that currently using their own impl as well. It would be great to support this in pytorch codebase natively.

Currently, pytorch doesn't support both python version and C++ API version of Adabelief optimizer. (though there are 3rd party libs on pip or you can implement your own with pytorch framework). I think we should add both to make pytorch consistent. 

For specifically C++ part, here is an example if you are interested working on it:
https://github.com/pytorch/pytorch/pull/40009

For python part, if you have your own impl already, do you mind sending a PR so we can review and adopt it into pytorch codebase?  We put python optimizer under torch/optim/

Let me know if you have any questions and I can assist here. Thanks.

**Comment by matovitch** [comment]
@glaringlee 
Thank you for your quick reply. For the python part, I have been using the PyPI package from the author of the paper. So sorry, no PR ready. I was proposing on helping on the C++ front because I am much more familiar with this language but I will aim for both Python and C++ and let you know if I need any assistance. Thank you for the PR example too.

**Comment by juntang-zhuang** [comment]
@glaringlee Thanks for your interest. I'm wondering what is the template or requirement to create a pull request to add AdaBelief to pytorch code base? We have an implementation (updated to version 0.2.0 already) in pip, just curious what are the requirements so that we can modify the original code for a merge. Thanks a lot.

**Comment by juntang-zhuang** [comment]
@matovitch Thanks a lot for your interest. I'm super glad to hear that AdaBelief helps with your experiments. We have release PyTorch and Tensorflow version 0.2.0 of AdaBelief in pip, but not C++ version yet since I'm not so familiar with C++. Really glad to hear you offer to help, please let me know if anything is needed from my end.

**Comment by glaringlee** [comment]
@juntang-zhuang 
Hi, we don't have any template written in a doc format since pytorch keeps changing soon. But here is a PR that adding another optimizer into pytorch c++ api. I think reading example might be a good way to go.
https://github.com/pytorch/pytorch/pull/40009

Let me know if you have any questions.

**Comment by matovitch** [comment]
Just to put an update. I did not have much time to work on it these past weeks. Sorry. I hope to be able to work on it in January (currently on a trip). Thank you @juntang-zhuang for letting me know of the new version (only default parameters change if I'm not mistaken). What needs to be done is clear, I just need to take some time to do it (if you don't want to wait @juntang-zhuang , you can try to do the Python part, and I could add the C++ part if you are not sure to be able to do it :))

**Comment by glaringlee** [comment]
@juntang-zhuang I updated my previous post with examples for both python and c++. Please take a look at.
@matovitch Thanks for updating status, please take your time.

**Comment by juntang-zhuang** [comment]
@glaringlee Thanks so much for the update. I'll take a careful look and create a pull request. 
@matovitch I will try to merge the python version first, since there's already implementations in pip. Thanks a lot for help with the C++ version. The update includes a fix of error in coupled weight decay (some error in adabelief-pytorch==0.1.0 causes the coupled weight decay not taking effect), and then some modification to default hyper-parameters (changes from 0.0.5 to 0.1.0, then 0.2.0 remains the same as 0.1.0). We have also updated the arxiv paper, and the updated algorithm part stresses the in-place add of eps to centered second momentum. Hopefully that makes the algorithm clear enough, please tell me if any clarification is needed. Thanks again for your interest.

**Comment by juntang-zhuang** [comment]
@glaringlee Hi, I was trying to implement the _multi_tensor version of AdaBelief, but I found the example code of adamw in PR #22355[https://github.com/pytorch/pytorch/pull/22355#issue-292982833] and PyTorch==1.7.1 seems to have different APIs, for example  torch._foreach_div_(A,B) accepts B as a scalar_list in PR, but in PyTorch1.7 we should use torch._foreach_div_scalar_list(A,B) where B is a scalar_list.  Another difference is  torch._foreach_add_(A,B,C), the PR seems to assume C is a list of float with the same length as A, but PyTorch1.7 assumes C is a scalar. So I'm wondering which version should I stick to? Thanks a lot.

**Comment by glaringlee** [comment]
@juntang-zhuang 
foreach_xxxx apis has some changes recently, please follow the apis in latest pytorch version. Thanks.

**Comment by juntang-zhuang** [comment]
@glaringlee Hi, I just submitted a pull request for a python API of AdaBelief, including normal optimizer and _multi_tensor optimizer. Please take a look if anything needs to be improved. Thanks a lot!

**Comment by juntang-zhuang** [comment]
@glaringlee Hi, I just noticed a few errors prompt after the pull request, some are for the html is not correctly built, others are related to line 166 of torch/optim/_multi_tensor/adabeleief.py, which is "max_exp_avg_var = torch._foreach_maximum(max_exp_avg_var, exp_avg_var)", and this line is modified from the _multi_tensor version of adam in the master branch. Is the api changed ? Or should I just ignore these errors? Thanks a lot.

**Comment by glaringlee** [comment]
@juntang-zhuang 
I am on vocation this week, have minimal web access, will try to take a look at within this week. If I cant't, will look at this 1st thing next Monday.

**Comment by albanD** [comment]
Hi @juntang-zhuang 

Small update on this as most of us where away on the past two weeks.
Unfortunately @glaringlee was not aware of the special policy we have for the python side of the torch.nn and torch.optim APIs as he is much more focused on the c++ APIs.
Sorry about the late reply here :/ 

In particular, you can see it in other similar feature requests [here for example](https://github.com/pytorch/pytorch/issues/48520#issuecomment-736602300).
The idea is that removing a function that was added is not possible and so we want to be very careful about what we add as it then takes more and more time to maintain the library and keep it high quality.
For that reason, we usually require that the corresponding function is "widely adopted by the community". This can be measured by different things but usually the number of citations on the corresponding paper is a good indication. There is no hard threshold and it can vary depending on the exact subfield, but for an optimizer like that I would expect several hundred citations at least.

For such new features, we recommend that you keep it in your own repo at the beginning and it can be moved to core in a couple month if there is a wide adoption.
Note that we are still happy to help you by adding any elementary op that you need into core if need be.

**Comment by juntang-zhuang** [comment]
@albanD Thanks a lot. I'll keep my local repository for general purpose use, will consider a pull request later.

**Comment by albanD** [comment]
That sounds great! Don't hesitate to ask in this issue or on the forum if you have any issue with implementing new optimizer/operators in pytorch!

You can link your local repo with the code here if you want. In case people looking for an implementation find this issue :)

**Comment by juntang-zhuang** [comment]
@albanD  Thanks a lot! The link to local repository with an implementation of AdaBelief in PyTorch is https://github.com/juntang-zhuang/Adabelief-Optimizer

**Comment by albanD** [comment]
Thanks for the link.

Closing for now to reduce noise. Please re-open the issue when we have more data on the adoption so that we can reconsider adding this!
